### DISCUSSION TITLE
BnF SRU Server

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#140] [Q&A] [answered]
with firefox I get a result
with BiblioteQ I can't translate in the bibliteq.conf file

the [SRU-2] for the BbnF server
it respects the version 1.2 of SRU according to the notice

[SRU-2]
name = Bibliothèque nationale de France
proxy_host =
proxy_password =
proxy_port =
proxy_type = None
proxy_username =
url_isbn = "https://catalogue.bnf.fr/api/SRU?version=1.2&operation=searchRetrieve&query=bib.isbn adj %1"

or 
url_isbn = "https://catalogue.bnf.fr/api/SRU?version=1.2&operation=searchRetrieve&query=bib.isbn adj "978-2-211-21125-3""

problem with quotation marks?

isbn-13 : 9782355848858

https://catalogue.bnf.fr/api/SRU?version=1.2&operation=searchRetrieve&query=bib.isbn adj "9782355848858"

**Comment by textbrowser** [comment, marked as answer]
Try url_isbn = "https://catalogue.bnf.fr/api/SRU?version=1.2&operation=searchRetrieve&query=bib.isbn adj '%1'".

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/commit/a732c4d1a59192a2817a2a47809b84cb8500183d

:)

**Reply by meteos77** [reply]
SRU Query Error
An SRU entry may not yet esist for the provided ISBN(s)

**Comment by textbrowser** [comment]
You have to reload biblioteq.conf or restart the program.

**Comment by textbrowser** [comment]
![Screenshot_20211209_085425](https://user-images.githubusercontent.com/10701156/145409370-48ed2747-42b7-441a-bc3e-3a85cf5a1567.png)

**Reply by meteos77** [reply]
same problem as with Z3950 - UTF8 / ISO8859-15

**Reply by meteos77** [reply]
it's ok with ISO8859-15

**Reply by meteos77** [reply]
the interpretation on the other hand does not work well
ex : 9791027607129
the title is not taken ...

there are 5 xml export formats

By default, the output format is unimarcXchange (Unimarc encapsulated in MarcXchange).

Unimarc format with analytics (encapsulated in MarcXchange) : "unimarcXchange-anl".
&recordSchema=unimarcXchange-anl

Intermarc format (encapsulated in MarcXchange)
&recordSchema=intermarcXchange

Intermarc format with analytics (encapsulated in MarcXchange) : "intermarcXchange-anl
&recordSchema=intermarcXchange-anl

Dublin Core format: "dublincore".
&recordSchema=dublincore

Translated with www.DeepL.com/Translator (free version)

**Comment by textbrowser** [comment]
Peruse biblioteq_marc.cc and see what needs to be added / fixed. :)

**Reply by meteos77** [reply]
all that concerns unimarc is not filled :-)
is MARC21 the format of the library of congress ?

**Comment by textbrowser** [comment]
My level of interest in parsing (and correcting the parsers) reached zero many years ago. :P

**Comment by textbrowser** [comment]
I have received numerous requests to correct / amend the parsers and whenever I do, something is broken for some obscure library or example or XML format.

**Comment by textbrowser** [comment]
I hope you're not upset with my emotional resistance to parsing. :)

**Comment by textbrowser** [comment]
I will look after correcting the new ticket.

**Reply by meteos77** [reply]
I understand and the Z3950 server works quite well for this functionality.
I was testing because the BnF had implemented the SRU it was an opportunity to do a real test because the library of congress never gave me any information about our books :-)

**Reply by textbrowser** [reply]
LoC is my test case. I cannot test every library. My assumption was that if it works for it, it'll work for others as long as they follow the syntax. It doesn't. So my assumption is incorrect or the formats are different.

**Reply by textbrowser** [reply]
People don't understand the complexities of library software. Or catalogs. It's an unclear frontier.

**Reply by meteos77** [reply]
there are a lot of standards and for me it's a lot of liners not too much evolution so the standards coexist for a very very long time
which doesn't facilitate development.
between 2 national libraries, they could respect the same version of standard or have a compatible standard during the transition.

**Reply by textbrowser** [reply]
Did I mention SQL? I did, right? I think I mentioned SQL. Now I'm mentioning SQL. :)

**Reply by textbrowser** [reply]
If you were able to connect to their database and extract the specific fields, it would be splendid. It's what their interface does. But they hide the databases through formats upon formats. It's quite a stupid solution because of security. Or something.

**Reply by textbrowser** [reply]
In BQ, you don't need BQ. You can write your own software to do whatever extraction and manipulation you need. BQ presents the data. That's it. Well, a little bit more than that.

**Comment by textbrowser** [comment]
I would prefer a library to parse these fields so I can simply issue results.value(field). The present solution requires that I maintain several complex parsers.

**Reply by meteos77** [reply]
I quickly read the BnF's notice and the version it uses is 1.2 ; the LoC may be using version 2.0, hence the problems
https://www.loc.gov/standards/sru/differences.html

the SRU service was just a test; 
the important record import format for me is unimarc by file because it is the one used by the Bibliothèque Départementale des Prêts

**Reply by meteos77** [reply]
something escapes me isn't the yaz library made to make your life easier ?
https://software.indexdata.com/yaz/doc/yaz.pdf

**Reply by meteos77** [reply]
Be carefull in mountain / hill mode :-)
another solution to avoid multiple parsing a multi-format converter to 1 single format and parsing

**Reply by meteos77** [reply]
https://sourceforge.net/projects/extractiso2709/

https://sourceforge.net/projects/fastmrclib/
summary : A small and simple library for reading and parsing MARC (bibliographic ISO-2709) records quickly. If you need to read and parse MRC files - it is that suits you. :

**Reply by textbrowser** [reply]
I cannot access the PDF. Site is down.

**Reply by meteos77** [reply]
no luck, it was "on" saturday
I went through the site https://www.indexdata.com/resources/software/yaz/

**Reply by textbrowser** [reply]
https://github.com/textbrowser/biblioteq/blob/master/Source/biblioteq_z3950results.cc

One has to parse the results. I would like a library that parses the results and provides easy access. Perhaps YAZ has changed since I integrated it.

**Reply by meteos77** [reply]
yaz - their site is up

**Reply by meteos77** [reply]
if you have a marc21 parser that works very well 

why not convert the unimarc format to marc21 and parse after the conversion?
because France uses unimarc but I saw that other countries can still have a different format the English -> ukmarc


The context of the project https://www.bl.uk/bibliographic/pdfs/usemarcon_manual.pdf

The different national MARC (MAchine Readable Catalogue) standards are seen as barriers to a wider exchange of bibliographic records in Europe and beyond. Worldwide, there are currently nearly 50 different MARC formats in use, 10 of which are used in national libraries in European Community countries. Such variation is a fundamental problem for libraries that wish to Such variation is a fundamental problem for libraries wishing to obtain or provide data internationally and often results in the re-cataloguing of materials for which records are readily available in other formats.

**Reply by textbrowser** [reply]
How do I convert UNIMARC to MARC21?

**Reply by meteos77** [reply]
https://github.com/datascouting/Unimarc2Marc21/blob/master/www/app/bash/unimarc2marc21.sh

yaz-marcdump

**Reply by textbrowser** [reply]
That's a script. That will not work in anything but Unix. :P

**Reply by meteos77** [reply]
I'm not the developer :-)
if yaz provides a command line tool to convert
unimarc to marc you can't integrate it in your program ?

https://github.com/indexdata/yaz/blob/master/util/marcdump.c

**Reply by meteos77** [reply]
I found this but the haskell language ???
https://github.com/Baksonator/marc-parser

**Reply by textbrowser** [reply]
> I'm not the developer :-) if yaz provides a command line tool to convert unimarc to marc you can't integrate it in your program ?
> 
> https://github.com/indexdata/yaz/blob/master/util/marcdump.c

That depends. Since it's a single file which requires the YAZ infrastructure, I have to test the integration on: FreeBSD, Linux, Mac, Windows. If those systems work, then I essentially have to babysit that file. I prefer a library. YAZ is integrated into BQ. Is this dump included in the library? On Android, YAZ is not available because it requires the large infrastructure. It could be but it isn't. So, YAZ is an optional library. SRU is the preferred mechanism. :) Or OpenLibrary.

**Reply by meteos77** [reply]
I am well aware of the difficulty with all the interlocking standards

after I give you my ideas that I meet during my research, it is surely bad tracks but the French sites do not necessarily give the same track as other countries so you never know.

**Reply by textbrowser** [reply]
You can also execute the dump yourself, paste the results in the MARC Tags and click the special button.

**Reply by meteos77** [reply]
I discovered another feature of BiblioteQ  Great

**Reply by meteos77** [reply]
I finally found where the word Guess was put in the interface :-)
I don't know how to translate it can you explain it to me other than by the word "Guess"?

**Reply by textbrowser** [reply]
That is why I requested an official document describing the UNIMARC format. :P

**Reply by textbrowser** [reply]
Estimate, Discover, etc.

**Reply by textbrowser** [reply]
Guess is short and the area where it's located is limited.

**Reply by meteos77** [reply]
https://www.loc.gov/marc/unimarctomarc21_intro.pdf

**Reply by meteos77** [reply]
it is neither unimarc , nor marc21

**Reply by meteos77** [reply]
this is not the norm but it suits you?
https://www.loc.gov/marc/unimarctomarc21.html

**Reply by textbrowser** [reply]
Only 100 pages?

**Reply by meteos77** [reply]
https://www.ifla.org/unimarc-updates/

**Reply by meteos77** [reply]
unimarc is only a derivative of the marc format (iso2709)

**Reply by meteos77** [reply]
https://www.ifla.org/wp-content/uploads/2019/05/assets/unimarc/unimarc_format_classification_concise.pdf

**Reply by textbrowser** [reply]
I acknowledge your willingness to help. :) I am interested in brief guides which correspond to things I've already parsed.
The MARC Tags are included in the databases as reminders to other customers. Things that they will take the time to invest in and place requests such as "I would like this field extracted (name of field, place, etc.) and placed in this new database field; here is the documentation which shows that the requested field is indeed what I think it is."

**Reply by textbrowser** [reply]
Requests such as "please parse UNIMARC" are... lazy?

**Reply by textbrowser** [reply]
OK, the concise one looks better.

**Reply by meteos77** [reply]
We didn't understand each other, I thought you were interested in the unimarc standard.
for me BiblioteQ, perfectly meets our needs concerning Z3950 and SRU

**Reply by meteos77** [reply]
my message from 3 days ago 
the Unimarc parser for Z3950 works very well it allows to retrieve the records without any problem

**Reply by textbrowser** [reply]
I didn't see this message. :)

**Reply by meteos77** [reply]
I'm going to sleep :-)

**Reply by textbrowser** [reply]
Are you at the library?

**Comment by meteos77** [comment]
by adding the format &recordSchema=intermarcXchange to the SRU line,
 I have the  Title OK ;  Authors OK ;Place of Publication OK  - Publisher KO

**Reply by textbrowser** [reply]
"Automatic" parsing? I will look at all of that again. Maybe it can be improved.

**Reply by meteos77** [reply]
if possible harmonize the version in 1.2 this will allow to detect problems on the 2 servers at the same time and do the work only once:-)

the dream is that everyone passes in 2.0 but ... in 2040 (I try to inquire at the Bnf)

**Comment by meteos77** [comment]
in biblioteq.conf add &recordSchema=intermarcXchange"

complete line :+1: 
url_isbn = "https://catalogue.bnf.fr/api/SRU?version=1.2&operation=searchRetrieve&query=bib.isbn adj '%1' &recordSchema=intermarcXchange"


the problem of encoding the configuration file is present
if UTF-8 the Bnf SRU does not work  (the SRU Bnf output is in UTF-8 according to their documentation)
if ISO8859-15 SRU Ok but then accented characters are incorrectly presented.


statement: your query to the Library of Congress SRU service is in version 1.1 (searchRetrieve&version=1.1)

**Reply by textbrowser** [reply]
Committed.

**Comment by textbrowser** [comment]
OK, I'm now looking at the SRU query.

**Reply by textbrowser** [reply]
Will look at this later. It can be improved.

**Comment by textbrowser** [comment]
Parsing of UNIMARC data for books is not implemented. :) That's a problem.

**Reply by meteos77** [reply]
the Unimarc parser for Z3950 works very well it allows you to retrieve the records without any problem

**Reply by textbrowser** [reply]
I made a change so that UNIMARC is parsed using MARC21 for SRU. If you can find the UNIMARC standard for SRU, I'll add it.

**Comment by meteos77** [comment]
I made a small assessment of Z39.50 and SRU with the French servers.

[bilan-Z3950-SRU-France.txt](https://github.com/textbrowser/biblioteq/files/7710719/bilan-Z3950-SRU-France.txt)

**for Z3950 Sudoc**: from what I observe the codes 0098 and 009C would be there to specify that the title contains an article which is not used for classification in certain cataloguing.
the 1st one is for example 
discreet charm of the intestine or  the discreet charm of the intestine

request : delete these 2 useless codes in the titles
the Sudoc server propagates the 10/10 documents


**for the Z3950 Bnf** : problem of end of the title field, you take the 200 $a (title) and 200 $b (printed text) 

request : remove the 200 $b from the title
the Bnf Z3950 server propagates the 9/10 documents


**for the SRU Bnf** : it's the one that works **the best :-)**
the BnF SRU server propagates the 9/10 documents

**Reply by textbrowser** [reply]
That's an interesting idea. Remove certain fields for certain servers?

**Reply by textbrowser** [reply]
I would introduce an ignore_fields setting in the biblioteq.conf file.

For example, Z39.50 Sudoc:

ignore_fields = "0098, 009C"

**Reply by textbrowser** [reply]
Thoughts? Notice that the values associated with ignore_fields are comma-separated. I think that's a valid delimiter.

**Reply by meteos77** [reply]
for me these "ascii code" are put to indicate that it is an article
"le" "la" "les" "l'" for you "a" "the"
 I don't think you have to remove the article but the codes
because with the other servers the title is well with the articles

**Reply by meteos77** [reply]
for you the migratory birds
0098the009cmigratory birds

if your proposal is we delete the 2 ascii codes (0098 and 009c)
it's great but we keep "the"

**Reply by textbrowser** [reply]
Can you paste a sample MARC Tags for each query?

**Reply by meteos77** [reply]
what sample ?

**Reply by textbrowser** [reply]
You performed a query via Z39.50 Sudoc. Paste the results of the query. :)

**Reply by meteos77** [reply]
txt format

**Comment by meteos77** [comment]
[9782330048815.txt](https://github.com/textbrowser/biblioteq/files/7713501/9782330048815.txt)

**Reply by meteos77** [reply]
[9782330048815.txt](https://github.com/textbrowser/biblioteq/files/7713541/9782330048815.txt)
full version

**Comment by meteos77** [comment]
02560cam0 2200529   450 
001 185219861
003 http://www.sudoc.fr/185219861
005 20210201142302.000
010    $a 978-2-330-04881-5 $b br. $d 21,80 EUR
035    $a (OCoLC)908129151
073  1 $a 9782330048815
100    $a 20150428h20152015m  y0frey50      ba
101 1  $a fre $c ger
102    $a FR
105    $a a   a   000yy
106    $a r
181    $6 z01 $c txt $2 rdacontent
181    $6 z02 $c sti $2 rdacontent
181  1 $6 z01 $a i# $b xxxe##
181  1 $6 z02 $a b# $b xb2e##
182    $6 z01 $c n $2 rdamedia
182    $6 z02 $c n $2 rdamedia
182  1 $6 z01 $a n
182  1 $6 z02 $a n
183  1 $6 z01 $a nga $2 RDAfrCarrier
200 1  $a Le charme discret de l'intestin $e tout sur un organe mal aimé $f Giulia Enders $g illustrations par Jill Enders $g traduit de l'allemand par Isabelle Liber
214  0 $a Arles $c Actes Sud $d DL 2015
215    $a 1 vol. (350 p.) $c ill., couv. ill. $d 22 cm

**Reply by textbrowser** [reply]
This example does not include 0098.

**Reply by textbrowser** [reply]
Nor 200 $b. :)

**Reply by meteos77** [reply]
200 1  $a Le charme discret de l'intestin

visible in txt format by github it replaces by �

**Reply by textbrowser** [reply]
OK, let me add the new option and you can test it to determine if it suits you.

**Reply by textbrowser** [reply]
This is the problem for you "$b - General Material Designation"?

**Reply by meteos77** [reply]
yes in SRU , it modifies the title
it should be in another field of the database if we want to keep it

**Reply by textbrowser** [reply]
Do you an example which includes 200 $b?

**Reply by meteos77** [reply]
[9782330048815-SRU.txt](https://github.com/textbrowser/biblioteq/files/7713636/9782330048815-SRU.txt)

**Reply by textbrowser** [reply]
That does not include 200. :) OK, I'll create an example myself and test.

The idea is that BQ will remove everything between $b to the next field, including $b. This is for Z39.50. We can look at SRU later.

I will not be able to complete this today. Tomorrow's the release.

**Reply by meteos77** [reply]
[9782330048815-Z3950-pb-200$b.txt](https://github.com/textbrowser/biblioteq/files/7713655/9782330048815-Z3950-pb-200.b.txt)

source Z3950 BnF server

**Reply by meteos77** [reply]
Anyway the BnF conf in Z39.50 is not included in the release so there will be no worries, I'm a step ahead :-)

**Reply by meteos77** [reply]
give all your energy in the output, my requests can wait

**Reply by meteos77** [reply]
for me the SRU server is the best ; the SRU file was to show you the difference
